File: plugin/flutter_oss_client-master/android/src/main/java/com/payne/flutter_client_oss/FOssClient.java
package com.payne.flutter_client_oss;

import android.annotation.SuppressLint;
import android.os.Handler;
import android.util.Log;

import com.alibaba.sdk.android.oss.ClientConfiguration;
import com.alibaba.sdk.android.oss.ClientException;
import com.alibaba.sdk.android.oss.OSS;
import com.alibaba.sdk.android.oss.OSSClient;
import com.alibaba.sdk.android.oss.ServiceException;
import com.alibaba.sdk.android.oss.callback.OSSCompletedCallback;
import com.alibaba.sdk.android.oss.common.auth.OSSCredentialProvider;
import com.alibaba.sdk.android.oss.common.auth.OSSStsTokenCredentialProvider;
import com.alibaba.sdk.android.oss.internal.OSSAsyncTask;
import com.alibaba.sdk.android.oss.model.OSSRequest;
import com.alibaba.sdk.android.oss.model.ResumableDownloadRequest;
import com.alibaba.sdk.android.oss.model.ResumableDownloadResult;
import com.alibaba.sdk.android.oss.model.ResumableUploadRequest;
import com.alibaba.sdk.android.oss.model.ResumableUploadResult;

import java.io.File;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Random;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class FOssClient {
    public int speedLimit = 0;
    private OSS mOss;
    final private MethodChannel channel;
    final Handler handler = new Handler();
    FOssClient(MethodChannel channel){
        this.channel = channel;
    }
    private Map<String, OSSAsyncTask> taskMap = new HashMap();
    public void initClient(MethodCall call,MethodChannel.Result result){
        Map<String,String> sts = call.argument("sts");
        String accessKeyId = sts.get("accessKeyId");
        String accessKeySecret = sts.get("accessKeySecret");
        String securityToken = sts.get("securityToken");
        String endPoint = call.argument("endPoint");
        // 从STS服务获取的临时访问密钥（AccessKey ID和AccessKey Secret）。
        if (accessKeyId == null || accessKeySecret == null || securityToken == null || endPoint == null){
            result.success(0);
            return;
        }
        OSSCredentialProvider provider = new OSSStsTokenCredentialProvider(accessKeyId,accessKeySecret,securityToken);
        ClientConfiguration conf = new ClientConfiguration();
        if (call.hasArgument("config")){
            Map<String,Integer> config = call.argument("config");
            Integer maxConcurrentRequest = config.get("maxConcurrentRequest");
            if (maxConcurrentRequest != null){
                conf.setMaxConcurrentRequest(maxConcurrentRequest);
            }
            Integer socketTimeout = config.get("socketTimeout");
            if (socketTimeout != null){
                conf.setSocketTimeout(socketTimeout);
            }
            Integer connectionTimeout = config.get("connectionTimeout");
            if (connectionTimeout != null){
                conf.setConnectionTimeout(connectionTimeout);
            }
            Integer maxLogSize = config.get("maxLogSize");
            if (maxLogSize != null){
                conf.setMaxLogSize(maxLogSize);
            }
            Integer maxErrorRetry = config.get("maxErrorRetry");
            if (maxErrorRetry != null){
                conf.setMaxErrorRetry(maxErrorRetry);
            }
        }
        mOss = new OSSClient(FlutterClientOssPlugin.weakActivity.get(),endPoint,provider,conf);
    }

    public void methodUpload(MethodCall call, MethodChannel.Result result){
        String bucketName = call.argument("bucketName");
        String ossFileName = call.argument("ossFileName");
        String filePath = call.argument("filePath");
        String recordDir = call.argument("recordDir");
        String id = createTaskId();
        asyncResumableUpload(bucketName,ossFileName,filePath,recordDir,id);
        result.success(id);
    }

    private void asyncResumableUpload(String bucketName,String ossFileName,String filePath,String recordDir,String taskId){
        Log.e("asyncResumableUpload","start-0");
        File recordDirFile = new File(recordDir);
        // 确保断点记录的保存文件夹已存在，如果不存在则新建断点记录的保存文件夹。
        if (!recordDirFile.exists()) {
            recordDirFile.mkdirs();
        }
        Log.e("asyncResumableUpload","start-1");
        ResumableUploadRequest request = new ResumableUploadRequest(bucketName,ossFileName,filePath,recordDir);
        request.setDeleteUploadOnCancelling(false);
        request.setThreadNum(1);
        request.setProgressCallback((request1, currentSize, totalSize) -> {
            Log.e("asyncResumableUpload","start-3");
            if (totalSize == 0) {
                Log.e("asyncResumableUpload","start-4");
                return;
            }
            Map<String, Object> args = new HashMap<>();
            args.put("current",currentSize);
            args.put("total",totalSize);
            args.put("taskId",taskId);
            Log.e("asyncResumableUpload","start-5");
            handler.post(() -> channel.invokeMethod("uploadProgress",args));
            Log.e("asyncResumableUpload","start-6");
        });
        OSSAsyncTask<ResumableUploadResult> task = mOss.asyncResumableUpload(request, new OSSCompletedCallback<ResumableUploadRequest, ResumableUploadResult>() {
            @Override
            public void onSuccess(ResumableUploadRequest request, ResumableUploadResult result) {
                Log.e("asyncResumableUpload","start-8");
                Map<String, Object> args = new HashMap<>();
                args.put("success",1);
                args.put("taskId",taskId);
                handler.post(() -> channel.invokeMethod("uploadResult",args));
            }

            @Override
            public void onFailure(ResumableUploadRequest request, ClientException clientException, ServiceException serviceException) {
                Map<String, Object> args = new HashMap<>();
                args.put("success",0);
                args.put("taskId",taskId);
                if (clientException != null) {
                    Log.e("asyncResumableUpload","start-9");
                    clientException.printStackTrace();
                    args.put("cancel", clientException.isCanceledException() ? 1 : 0);
                    args.put("message", clientException.getMessage());
                    handler.post(() -> channel.invokeMethod("uploadResult", args));
                } else if (serviceException != null) {
                    Log.e("asyncResumableUpload","start-10");
                    args.put("message",serviceException.getErrorCode());
                    args.put("cancel",0);
                    handler.post(() -> channel.invokeMethod("uploadResult",args));
                    serviceException.printStackTrace();
                }
            }
        });
        Log.e("asyncResumableUpload","start-2");
        taskMap.put(taskId,task);
    }

    public void methodDownload(MethodCall call, MethodChannel.Result result){
        String bucketName = call.argument("bucketName");
        String ossFileName = call.argument("ossFileName");
        String filePath = call.argument("filePath");
        String recordFile = call.argument("recordFile");
        Log.e("methodDownload",bucketName + "--" + ossFileName + "--" + filePath +  "--" + recordFile);
        String id = createTaskId();
        asyncResumableDownload(bucketName,ossFileName,filePath,recordFile,id);
        result.success(id);
    }

    public void asyncResumableDownload(String bucketName, String ossFileName, String filePath, String recordFile,String taskId) {

        ResumableDownloadRequest request = new ResumableDownloadRequest(bucketName, ossFileName, filePath);
        request.setEnableCheckPoint(true);
        request.setCheckPointFilePath(recordFile);
        request.setThreadNum(1);
        if (speedLimit > 0) {
            Map<String, String> headers = new HashMap<>();
            headers.put("x-oss-traffic-limit",speedLimit + "");
            request.setRequestHeader(headers);
        }
        request.setProgressListener((request1, currentSize, totalSize) -> {
            if (totalSize == 0) {
                return;
            }
            Map<String, Object> args = new HashMap<>();
            args.put("current",currentSize);
            args.put("total",totalSize);
            args.put("taskId",taskId);
            handler.post(() -> channel.invokeMethod("downloadProgress",args));
        });
        final OSSAsyncTask<ResumableDownloadResult> task = mOss.asyncResumableDownload(request, new OSSCompletedCallback<ResumableDownloadRequest, ResumableDownloadResult>() {
            @Override
            public void onSuccess(ResumableDownloadRequest request, ResumableDownloadResult result) {
                Map<String, Object> args = new HashMap<>();
                args.put("success",1);
                args.put("taskId",taskId);
                handler.post(() -> channel.invokeMethod("downloadResult",args));
            }

            @Override
            public void onFailure(ResumableDownloadRequest request, ClientException clientException, ServiceException serviceException) {
                Map<String, Object> args = new HashMap<>();
                args.put("success",0);
                args.put("taskId",taskId);
                if (clientException != null) {
                    if (taskMap.get(taskId) == null) { //这个代表用户主动取消，去执行回调
                        args.put("cancel", clientException.isCanceledException() ? 1 : 0);
                        args.put("message", clientException.getMessage());
                        handler.post(() -> channel.invokeMethod("downloadResult", args));
                    }
                    clientException.printStackTrace();;
                } else if (serviceException != null) {
                    args.put("message",serviceException.getErrorCode());
                    args.put("cancel",0);
                    handler.post(() -> channel.invokeMethod("downloadResult",args));
                    serviceException.printStackTrace();
                }
            }
        });
        taskMap.put(taskId,task);
    }

    public void checkFile(MethodCall call, MethodChannel.Result result){
        String bucketName = call.argument("bucketName");
        String fileName = call.argument("fileName");
        try {
           boolean r = mOss.doesObjectExist(bucketName, fileName);
           result.success(r ? 1 : 0);
        }catch (Exception e){
            result.success(0);
        }
    }

    public void getSignUrl(MethodCall call, MethodChannel.Result result){
        String bucketName = call.argument("bucketName");
        String fileName = call.argument("fileName");
       String r = mOss.presignPublicObjectURL(bucketName,fileName);
       result.success(r);
    }

    public void cancelTask(String taskId){
        OSSAsyncTask task = taskMap.get(taskId);
        if (task != null && !task.isCanceled() && !task.isCompleted()) {
            taskMap.remove(taskId);
            task.cancel();
        }
    }

    public void cancelAllTask(){
        for (String key : taskMap.keySet()) {
            OSSAsyncTask task = taskMap.get(key);
            if (task != null && !task.isCanceled() && !task.isCompleted()) {
                taskMap.remove(key);
                task.cancel();
            }
        }
        taskMap.clear();
    }

    @SuppressLint("DefaultLocale")
    private String createTaskId(){
        StringBuilder randomStr = new StringBuilder();
        for (int i = 0; i < 4; i++) {
            randomStr.append(String.format("%02d",(new Random()).nextInt(10000) % 100));
        }
        return System.currentTimeMillis() + randomStr.toString();
    }
}

File: plugin/flutter_oss_client-master/ios/Classes/FOssClient.swift
//
//  FOssClient.swift
//  flutter_client_oss
//
//  Created by lv on 2024/10/12.
//

import UIKit
import Flutter
import AliyunOSSiOS

class FOssClient: NSObject {
    var speedLimit:Int?
    private var client: OSSClient!
    private lazy var taskMap = [String:FOssClientTask]()
    private let channel:FlutterMethodChannel
    
    init(channel: FlutterMethodChannel) {
        self.channel = channel
    }
    
    func initSdk(_ call: FlutterMethodCall,result: @escaping FlutterResult){
        guard let args = call.arguments as? [String: Any?] else {
            result(0)
            return
        }
        guard let sts = args["sts"] as? [String: Any?] else {
            result(0)
            return
        }
        guard let accessKeyId = sts["accessKeyId"] as? String else {
            result(0)
            return
        }
        guard let accessKeySecret = sts["accessKeySecret"] as? String else {
            result(0)
            return
        }
        guard let securityToken = sts["securityToken"] as? String else {
            result(0)
            return
        }
        guard let endPoint = args["endPoint"] as? String else {
            result(0)
            return
        }
        let provider = OSSStsTokenCredentialProvider.init(accessKeyId: accessKeyId, secretKeyId: accessKeySecret, securityToken: securityToken)
        client = OSSClient.init(endpoint: endPoint, credentialProvider: provider)
        print("=================initSdk")
        print(client)
        result(1)
    }
    
    // func asyncResumableUpload(_ call: FlutterMethodCall,result: @escaping FlutterResult){
    //     guard let args = call.arguments as? [String: Any?] else {
    //         result(0)
    //         return
    //     }
    //     guard let bucketName = args["bucketName"] as? String else {
    //         result(0)
    //         return
    //     }
    //     guard let ossFileName = args["ossFileName"] as? String else {
    //         result(0)
    //         return
    //     }
    //     guard let filePath = args["filePath"] as? String else {
    //         result(0)
    //         return
    //     }
    //     guard let recordDir = args["recordDir"] as? String else {
    //         result(0)
    //         return
    //     }
    //     let managerTask = FOssClientTask()
    //     let taskId = "\(managerTask.hash)"
    //     let upload = OSSResumableUploadRequest()
    //     upload.bucketName = bucketName
    //     upload.objectKey = ossFileName
    //     var isDir: ObjCBool = false
    //     if !FileManager.default.fileExists(atPath: recordDir, isDirectory: &isDir) {
    //         try? FileManager.default.createDirectory(atPath: recordDir, withIntermediateDirectories: true)
    //     }
    //     upload.recordDirectoryPath = recordDir
    //     upload.deleteUploadIdOnCancelling = false
    //     upload.uploadingFileURL = NSURL.fileURL(withPath: filePath)
    //     upload.uploadProgress = {[weak self] send,totalBytes,totalBytesExpected in
    //         if (totalBytesExpected == 0){
    //             return
    //         }
    //         DispatchQueue.main.async {
    //             self?.channel.invokeMethod("uploadProgress", arguments: [
    //                 "current": totalBytes,
    //                 "total": totalBytesExpected,
    //                 "taskId": taskId
    //             ])
    //         }
    //     }
    //     let resumeTask = client.resumableUpload(upload)
    //     resumeTask.continue({[weak self] task in
    //         if let error = task.error as? NSError {
    //             if (error.code == 5){
    //                 //用户取消
    //                 DispatchQueue.main.async {
    //                     self?.channel.invokeMethod("uploadResult", arguments: [
    //                         "success":0,
    //                         "taskId": taskId,
    //                         "cancel":1,
    //                         "message": error.localizedDescription
    //                     ])
    //                 }
    //             }else{
    //                 DispatchQueue.main.async {
    //                     self?.channel.invokeMethod("uploadResult", arguments: [
    //                         "success":0,
    //                         "taskId": taskId,
    //                         "cancel":0,
    //                         "message": error.localizedDescription
    //                     ])
    //                 }
    //             }
    //         }else{
    //             DispatchQueue.main.async {
    //                 self?.channel.invokeMethod("uploadResult", arguments: [
    //                     "success":1,
    //                     "taskId": taskId
    //                 ])
    //             }
    //         }
    //         return nil
    //     })
    //     managerTask.task = resumeTask
    //     managerTask.request = upload
    //     taskMap[taskId] = managerTask
    //     result(taskId)
    // }

    func asyncResumableUpload(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        // 参数验证部分保持不变
        guard let args = call.arguments as? [String: Any?],
            let bucketName = args["bucketName"] as? String,
            let ossFileName = args["ossFileName"] as? String,
            let filePath = args["filePath"] as? String,
            let recordDir = args["recordDir"] as? String else {
            result(0)
            return
        }
        
        // 使用更高优先级的后台队列处理上传任务
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            
            // 文件检查部分保持不变
            guard let fileAttributes = try? FileManager.default.attributesOfItem(atPath: filePath),
                let fileSize = fileAttributes[.size] as? UInt64 else {
                DispatchQueue.main.async {
                    result(0)
                }
                return
            }
            
            let managerTask = FOssClientTask()
            let taskId = "\(managerTask.hash)"
            let upload = OSSResumableUploadRequest()
            
            // 基本配置保持不变
            upload.bucketName = bucketName
            upload.objectKey = ossFileName
            upload.recordDirectoryPath = recordDir
            upload.deleteUploadIdOnCancelling = false
            upload.uploadingFileURL = NSURL.fileURL(withPath: filePath)
            
            // 优化分片大小策略
            if fileSize > 500 * 1024 * 1024 { // 大于500MB
                upload.partSize = 2 * 1024 * 1024  // 2MB分片
            } else if fileSize > 100 * 1024 * 1024 { // 大于100MB
                upload.partSize = 1024 * 1024  // 1MB分片
            } else if fileSize > 10 * 1024 * 1024 { // 大于10MB
                upload.partSize = 512 * 1024  // 512KB分片
            } else {
                upload.partSize = 256 * 1024  // 256KB分片
            }
            
            // 进度回调优化
            var lastProgressTime = Date()
            var lastProgress: Float = 0
            let progressQueue = DispatchQueue(label: "com.oss.progress")
            
            // 根据文件大小动态调整进度回调策略
            let minProgressInterval: TimeInterval
            let minProgressChange: Float
            
            if fileSize > 100 * 1024 * 1024 {
                minProgressInterval = 1.0     // 大文件每秒最多更新一次
                minProgressChange = 0.02      // 至少2%的变化
            } else if fileSize > 10 * 1024 * 1024 {
                minProgressInterval = 0.5     // 中等文件每0.5秒最多更新一次
                minProgressChange = 0.01      // 至少1%的变化
            } else {
                minProgressInterval = 0.2     // 小文件每0.2秒最多更新一次
                minProgressChange = 0.005     // 至少0.5%的变化
            }
            
            upload.uploadProgress = { [weak self] send, totalBytes, totalBytesExpected in
                guard let self = self,
                    totalBytesExpected > 0 else { return }
                
                let currentProgress = Float(totalBytes) / Float(totalBytesExpected)
                let now = Date()
                
                // 使用专门的队列处理进度逻辑
                progressQueue.async {
                    // 检查是否满足更新条件
                    if now.timeIntervalSince(lastProgressTime) >= minProgressInterval &&
                    abs(currentProgress - lastProgress) >= minProgressChange {
                        
                        lastProgressTime = now
                        lastProgress = currentProgress
                        
                        DispatchQueue.main.async {
                            self.channel.invokeMethod("uploadProgress", arguments: [
                                "current": totalBytes,
                                "total": totalBytesExpected,
                                "taskId": taskId
                            ])
                        }
                    }
                }
            }
            
            // 创建记录目录
            var isDir: ObjCBool = false
            if !FileManager.default.fileExists(atPath: recordDir, isDirectory: &isDir) {
                try? FileManager.default.createDirectory(atPath: recordDir, withIntermediateDirectories: true)
            }
            
            // 执行上传
            let resumeTask = self.client.resumableUpload(upload)
            
            // 处理上传结果
            resumeTask.continue({ [weak self] task in
                guard let self = self else { return nil }
                
                // 确保在主线程处理结果
                DispatchQueue.main.async {
                    // 最后一次进度更新
                    if let result = task.result as? OSSResumableUploadResult {
                        self.channel.invokeMethod("uploadProgress", arguments: [
                            "current": fileSize,
                            "total": fileSize,
                            "taskId": taskId
                        ])
                    }
                    
                    if let error = task.error as? NSError {
                        if error.code == 5 {
                            self.channel.invokeMethod("uploadResult", arguments: [
                                "success": 0,
                                "taskId": taskId,
                                "cancel": 1,
                                "message": error.localizedDescription
                            ])
                        } else {
                            self.channel.invokeMethod("uploadResult", arguments: [
                                "success": 0,
                                "taskId": taskId,
                                "cancel": 0,
                                "message": error.localizedDescription
                            ])
                        }
                    } else {
                        self.channel.invokeMethod("uploadResult", arguments: [
                            "success": 1,
                            "taskId": taskId
                        ])
                    }
                }
                return nil
            })
            
            // 保存任务引用
            managerTask.task = resumeTask
            managerTask.request = upload
            self.taskMap[taskId] = managerTask
            
            // 返回任务ID
            DispatchQueue.main.async {
                result(taskId)
            }
        }
    }
    
    func asyncResumableDownload(_ call: FlutterMethodCall,result: @escaping FlutterResult){
        guard let args = call.arguments as? [String: Any?] else {
            result(0)
            return
        }
        guard let bucketName = args["bucketName"] as? String else {
            result(0)
            return
        }
        guard let ossFileName = args["ossFileName"] as? String else {
            result(0)
            return
        }
        guard let filePath = args["filePath"] as? String else {
            result(0)
            return
        }
        guard let recordDir = args["recordFile"] as? String else {
            result(0)
            return
        }
        
        let managerTask = FOssClientTask()
        let taskId = "\(managerTask.hash)"
        taskMap[taskId] = managerTask
        DispatchQueue.global().async {[weak self] in
            var isDir: ObjCBool = false
            if !FileManager.default.fileExists(atPath: recordDir, isDirectory: &isDir) {
                try? FileManager.default.createDirectory(atPath: recordDir, withIntermediateDirectories: true)
            }
            let headObject = self?.getFileInfo(bucketName: bucketName, objectKey: ossFileName)
            if headObject == nil{
                DispatchQueue.main.async {
                    self?.channel.invokeMethod("downloadResult", arguments: [
                        "success":0,
                        "taskId": taskId,
                        "cancel":0,
                        "message": "获取元数据失败"
                    ])
                }
                return
            }
            print(headObject!.objectMeta)
            let md5 = headObject!.objectMeta["ETag"]
            let totalSize:Int64 = Int64((headObject!.objectMeta["Content-Length"] as? String) ?? "0") ?? 0
            //判断记录文件是否存在
            let recordFile = "\(recordDir)/\(String(describing: md5))"
            if !FileManager.default.fileExists(atPath: recordFile, isDirectory: &isDir) {
                FileManager.default.createFile(atPath: recordFile, contents: nil)
            }
            //获取本地文件大小
            if !FileManager.default.fileExists(atPath: filePath, isDirectory: &isDir) {
                FileManager.default.createFile(atPath: filePath, contents: nil)
            }
            let attrs = try? FileManager.default.attributesOfItem(atPath: filePath)
            if attrs == nil {
                return
            }
            print(attrs!)
            let fileSize:Int64 = (attrs![FileAttributeKey.size] as? Int64) ?? 0
            print(totalSize)
            print(fileSize)
            if (fileSize >= totalSize){
                result(taskId)
                DispatchQueue.main.async {
                    self?.channel.invokeMethod("downloadProgress", arguments: [
                        "current": totalSize,
                        "total": totalSize,
                        "taskId": taskId
                    ])
                    self?.channel.invokeMethod("downloadResult", arguments: [
                        "success":1,
                        "taskId": taskId
                    ])
                }
                return
            }
            var headMap = [String:String]()
            let request = OSSGetObjectRequest()
            if let limit = self?.speedLimit {
                if (limit > 0){
                    headMap["x-oss-traffic-limit"] = "\(limit)"
                }
            }
            headMap["Range"] = "bytes=\(fileSize)-"
            print(headMap)
            request.headerFields = headMap
            request.bucketName = bucketName
            request.objectKey = ossFileName
            request.onRecieveData = { data in
                let handler = FileHandle.init(forWritingAtPath: filePath)
                if #available(iOS 13.4, *) {
                    try? handler?.seekToEnd()
                    handler?.write(data)
                    try? handler?.close()
                } else {
                    // Fallback on earlier versions
                }
                
            }
            // 可选字段。
            request.downloadProgress = {[weak self] bytesWritten, totalBytesWritten,totalBytesExpectedToWrite in
                // 当前下载段长度、当前已经下载总长度、一共需要下载的总长度。
                if (totalBytesExpectedToWrite == 0){
                    return
                }
                DispatchQueue.main.async {
                    self?.channel.invokeMethod("downloadProgress", arguments: [
                        "current": totalBytesWritten + fileSize,
                        "total": totalSize,
                        "taskId": taskId
                    ])
                }
            };
            let getTask = self?.client.getObject(request)
            getTask?.continue({[weak self] task in
                if let error = task.error as? NSError {
                    if (error.code == 5){
                        //用户取消
                        DispatchQueue.main.async {
                            self?.channel.invokeMethod("downloadResult", arguments: [
                                "success":0,
                                "taskId": taskId,
                                "cancel":1,
                                "message": error.localizedDescription
                            ])
                        }
                    }else{
                        DispatchQueue.main.async {
                            self?.channel.invokeMethod("downloadResult", arguments: [
                                "success":0,
                                "taskId": taskId,
                                "cancel":0,
                                "message": error.localizedDescription
                            ])
                        }
                    }
                }else{
                    DispatchQueue.main.async {
                        self?.channel.invokeMethod("downloadResult", arguments: [
                            "success":1,
                            "taskId": taskId
                        ])
                    }
                }
                return nil
            })
            managerTask.task = getTask
            managerTask.request = request
            result(taskId)
        }
        
    }
    
    func checkFile(_ call: FlutterMethodCall,result: @escaping FlutterResult){
        
    }
    
    func getSignUrl(_ call: FlutterMethodCall,result: @escaping FlutterResult){
        
    }
    
    func cancelTask(taskId: String) {
        if let managerTask = taskMap[taskId],
           let ossTask = managerTask.task {
            if !ossTask.isCancelled && !ossTask.isCompleted {
                managerTask.request?.cancel()
                ossTask.waitUntilFinished() // 等待任务完全取消
                taskMap.removeValue(forKey: taskId)
            }
        }
    }
    
    func cancelAllTask() {
        taskMap.values.forEach { task in
            if let ossTask = task.task,
               !ossTask.isCancelled && !ossTask.isCompleted {
                task.request?.cancel()
                ossTask.waitUntilFinished() // 等待每个任务完全取消
            }
        }
        taskMap.removeAll()
    }
    
    private func getFileInfo(bucketName:String,objectKey:String)-> OSSHeadObjectResult?{
        let request = OSSHeadObjectRequest()
        request.bucketName = bucketName
        request.objectKey = objectKey
        let task = client.headObject(request)
        var headResult: OSSHeadObjectResult?
        task.continue({ t in
            if (task.error == nil){
                headResult = task.result as? OSSHeadObjectResult
            }
            return nil
        })
        task.waitUntilFinished()
        return headResult
    }

    
//    public void checkFile(MethodCall call, MethodChannel.Result result){
//           String bucketName = call.argument("bucketName");
//           String fileName = call.argument("fileName");
//           try {
//              boolean r = mOss.doesObjectExist(bucketName, fileName);
//              result.success(r ? 1 : 0);
//           }catch (Exception e){
//               result.success(0);
//           }
//       }
//
//       public void getSignUrl(MethodCall call, MethodChannel.Result result){
//           String bucketName = call.argument("bucketName");
//           String fileName = call.argument("fileName");
//          String r = mOss.presignPublicObjectURL(bucketName,fileName);
//          result.success(r);
//       }
//
//       public void cancelTask(String taskId){
//           OSSAsyncTask task = taskMap.get(taskId);
//           if (task != null && !task.isCanceled() && !task.isCompleted()) {
//               taskMap.remove(taskId);
//               task.cancel();
//           }
//       }
//
//       public void cancelAllTask(){
//           for (String key : taskMap.keySet()) {
//               OSSAsyncTask task = taskMap.get(key);
//               if (task != null && !task.isCanceled() && !task.isCompleted()) {
//                   taskMap.remove(key);
//                   task.cancel();
//               }
//           }
//           taskMap.clear();
//       }
//
//       @SuppressLint("DefaultLocale")
//       private String createTaskId(){
//           StringBuilder randomStr = new StringBuilder();
//           for (int i = 0; i < 4; i++) {
//               randomStr.append(String.format("%02d",(new Random()).nextInt(10000) % 100));
//           }
//           return System.currentTimeMillis() + randomStr.toString();
//       }
}
